pkg/rc/replication_controller.go

```diff
@@ -633,81 +661,267 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
-	rc.mu.Lock()
-	newNodes, err := rc.scheduler.AllocateNodes(rc.Manifest, rc.NodeSelector, nodesRequested)
-	rc.mu.Unlock()
+	newNodes, err := rc.scheduler.AllocateNodes(man, sel, nodesRequested)
 	if err != nil || len(newNodes) < 1 {
 		errMsg := fmt.Sprintf("Unable to allocate nodes over grpc: %s", err)
 		err := rc.alerter.Alert(rc.alertInfo(errMsg), alerting.LowUrgency)
 		if err != nil {
 			rc.logger.WithError(err).Errorln("Unable to send alert")
 		}
 
-		return util.Errorf(errMsg)
+		return "", util.Errorf(errMsg)
 	}
 
-	var modifyIndex uint64
-	if statusErr != nil && statusstore.IsNoStatus(statusErr) {
-		modifyIndex = 0
-	} else {
-		modifyIndex = queryMeta.LastIndex
-	}
+	newNode := newNodes[0]
 
-	status.NodeTransfer = &rcstatus.NodeTransfer{
-		// Again, we only support one node transfer at a time so just take the first old and new
-		OldNode: ineligibleCurrent[0],
-		NewNode: newNodes[0],
+	status := rcstatus.Status{
+		NodeTransfer: &rcstatus.NodeTransfer{
+			OldNode: oldNode,
+			NewNode: newNode,
+		},
 	}
+
 	writeCtx, writeCancel := transaction.New(context.Background())
 	defer writeCancel()
-	err = rc.rcStatusStore.CASTxn(writeCtx, rc.ID(), modifyIndex, status)
+	err = rc.rcStatusStore.CASTxn(writeCtx, rc.ID(), 0, status)
 	if err != nil {
-		return util.Errorf("Could not write new node to store: %s", err)
+		return "", util.Errorf("Could not write new node to store: %s", err)
 	}
 
 	err = transaction.MustCommit(writeCtx, rc.txner)
 	if err != nil {
-		return util.Errorf("Could not commit CASTxn: %s", err)
+		return "", util.Errorf("Could not commit CASTxn: %s", err)
+	}
+
+	return newNode, nil
+}
+
+func (rc *replicationController) scheduleWithSession(newNode types.NodeName) error {
+	rc.logger.NoFields().Infof("Scheduling on %s", newNode)
+	rc.mu.Lock()
+	manifest := rc.Manifest
+	rc.mu.Unlock()
+
+	manifestBytes, err := manifest.Marshal()
+	if err != nil {
+		return err
+	}
+
+	key := path.Join(string(consul.INTENT_TREE), newNode.String(), manifest.ID().String())
+	name := fmt.Sprintf("wait-on-%s-health", newNode)
+
+	ctx, ctxCancel := context.WithCancel(context.Background())
+	_, session, err := consul.SessionContext(ctx, rc.consulClient, name)
+	if err != nil {
+		return err
+	}
+
+	rc.nodeTransfer.session = session
+
+	// Call this function to clean up the session when the node
+	// transfer finishes or is stopped
+	rc.nodeTransfer.sessionCancel = ctxCancel
+
+	writeCtx, writeCancel := transaction.New(context.Background())
+	defer writeCancel()
+
+	txnUnlocker, err := session.LockIfKeyNotExistsTxn(writeCtx, key, manifestBytes)
+	if err != nil {
+		return nil
+	}
+	rc.nodeTransfer.unlocker = txnUnlocker
+
+	ok, resp, err := transaction.Commit(writeCtx, rc.txner)
+	switch {
+	case err != nil:
+		return err
+	case !ok:
+		return util.Errorf("could not schedule new node due to a transaction violation: %s", transaction.TxnErrorsToString(resp.Errors))
```

Comment:
I think here is probably the right place to `return nil` actually and just log the error along with a message like "new RC likely scheduled this node already" or something. The reason is that this function has other cases where it will return an error so we don't want to just assume everything went okay in the step above this

Reply:
if we don't return an error here the calling function won't realize that this RC failed to schedule a new node. It will keep executing and spin up the goroutine and everything

Reply:
oh I see. In that case keep it the way it is